src/intra_edge.rs

```diff
@@ -4,19 +4,63 @@ use crate::src::levels::BL_128X128;
 use crate::src::levels::BL_16X16;
 use crate::src::levels::BL_32X32;
 use crate::src::levels::BL_64X64;
+use bitflags::bitflags;
+use std::ops::Shr;
+
+bitflags! {
+    #[derive(Clone, Copy, PartialEq, Eq, Hash, Default)]
+    pub(crate) struct EdgeFlags: u8 {
+        const I444_TOP_HAS_RIGHT = 1 << 0;
+        const I422_TOP_HAS_RIGHT = 1 << 1;
+        const I420_TOP_HAS_RIGHT = 1 << 2;
+
+        const I444_LEFT_HAS_BOTTOM = 1 << 3;
+        const I422_LEFT_HAS_BOTTOM = 1 << 4;
+        const I420_LEFT_HAS_BOTTOM = 1 << 5;
+    }
+}
+
+impl EdgeFlags {
+    const LEFT_HAS_BOTTOM: Self = Self::union_all([
+        Self::I444_LEFT_HAS_BOTTOM,
+        Self::I422_LEFT_HAS_BOTTOM,
+        Self::I420_LEFT_HAS_BOTTOM,
+    ]);
+
+    const TOP_HAS_RIGHT: Self = Self::union_all([
+        Self::I444_TOP_HAS_RIGHT,
+        Self::I422_TOP_HAS_RIGHT,
+        Self::I420_TOP_HAS_RIGHT,
+    ]);
+
+    pub const fn union_all<const N: usize>(flags: [Self; N]) -> Self {
+        let mut i = 0;
+        let mut output = Self::empty();
 
-pub type EdgeFlags = u8;
-pub const EDGE_I420_LEFT_HAS_BOTTOM: EdgeFlags = 32;
-pub const EDGE_I422_LEFT_HAS_BOTTOM: EdgeFlags = 16;
-pub const EDGE_I444_LEFT_HAS_BOTTOM: EdgeFlags = 8;
-pub const EDGE_I420_TOP_HAS_RIGHT: EdgeFlags = 4;
-pub const EDGE_I422_TOP_HAS_RIGHT: EdgeFlags = 2;
-pub const EDGE_I444_TOP_HAS_RIGHT: EdgeFlags = 1;
+        while i < N {
+            output = output.union(flags[i]);
+            i += 1;
+        }
 
-pub const EDGE_LEFT_HAS_BOTTOM: EdgeFlags =
-    EDGE_I444_LEFT_HAS_BOTTOM | EDGE_I422_LEFT_HAS_BOTTOM | EDGE_I420_LEFT_HAS_BOTTOM;
-pub const EDGE_TOP_HAS_RIGHT: EdgeFlags =
-    EDGE_I444_TOP_HAS_RIGHT | EDGE_I422_TOP_HAS_RIGHT | EDGE_I420_TOP_HAS_RIGHT;
+        output
+    }
+
+    pub(crate) const fn select(&self, select: bool) -> Self {
+        if select {
+            *self
+        } else {
+            Self::empty()
+        }
+    }
+}
+
+impl Shr<u32> for EdgeFlags {
```

Comment:
Could you `impl` this for `Rav1dPixelLayout` directly?  Instead of `u32` I mean.  I think the shift makes more sense operating on both types rather than a more arbitrary `u32`.